flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/StatefulSequenceSource.java

```diff
@@ -61,52 +65,89 @@
 			}
 		} else {
-			// the first time the job is executed
-
-			final int stepSize = getRuntimeContext().getNumberOfParallelSubtasks();
 			final int taskIdx = getRuntimeContext().getIndexOfThisSubtask();
-			final long congruence = start + taskIdx;
+			final int parallelTasks = getRuntimeContext().getNumberOfParallelSubtasks();
+
+			final long totalElements = Math.abs(end - start + 1L);
+			final int maxParallelism = getRuntimeContext().getMaxNumberOfParallelSubtasks();
+			final int totalPartitions = totalElements < Integer.MAX_VALUE ? Math.min(maxParallelism, (int) totalElements) : maxParallelism;
 
-			long totalNoOfElements = Math.abs(end - start + 1);
-			final int baseSize = safeDivide(totalNoOfElements, stepSize);
-			final int toCollect = (totalNoOfElements % stepSize > taskIdx) ? baseSize + 1 : baseSize;
+			Tuple2<Integer, Integer> localPartitionRange = getLocalRange(totalPartitions, parallelTasks, taskIdx);
+			int localStartIdx = localPartitionRange.f0;
+			int localEndIdx = localStartIdx + localPartitionRange.f1;
 
-			for (long collected = 0; collected < toCollect; collected++) {
-				this.valuesToEmit.add(collected * stepSize + congruence);
+			for (int partIdx = localStartIdx; partIdx < localEndIdx; partIdx++) {
+				Tuple2<Long, Long> limits = getPartitionLimits(totalElements, totalPartitions, partIdx);
+				endToNextOffsetMapping.put(limits.f1, limits.f0);
 			}
 		}
 	}
 
+	private Tuple2<Integer, Integer> getLocalRange(int totalPartitions, int parallelTasks, int taskIdx) {
+		int minPartitionSliceSize = totalPartitions / parallelTasks;
+		int remainingPartitions = totalPartitions - minPartitionSliceSize * parallelTasks;
+
+		int localRangeStartIdx = taskIdx * minPartitionSliceSize + Math.min(taskIdx, remainingPartitions);
+		int localRangeSize = taskIdx < remainingPartitions ? minPartitionSliceSize + 1 : minPartitionSliceSize;
+
+		return new Tuple2<>(localRangeStartIdx, localRangeSize);
+	}
+
+	private Tuple2<Long, Long> getPartitionLimits(long totalElements, int totalPartitions, long partitionIdx) {
+		long minElementPartitionSize = totalElements / totalPartitions;
+		long remainingElements = totalElements - minElementPartitionSize * totalPartitions;
+		long startOffset = start;
+
+		for (int idx = 0; idx < partitionIdx; idx++) {
+			long partitionSize = idx < remainingElements ? minElementPartitionSize + 1L : minElementPartitionSize;
+			startOffset += partitionSize;
+		}
+
+		long partitionSize = partitionIdx < remainingElements ? minElementPartitionSize + 1L : minElementPartitionSize;
+		return new Tuple2<>(startOffset, startOffset + partitionSize);
+	}
+
 	@Override
 	public void run(SourceContext<Long> ctx) throws Exception {
-		while (isRunning && !this.valuesToEmit.isEmpty()) {
-			synchronized (ctx.getCheckpointLock()) {
-				ctx.collect(this.valuesToEmit.poll());
+		for (Map.Entry<Long, Long> partition: endToNextOffsetMapping.entrySet()) {
```

Comment:
I wonder if it would make sense to emit local ranges by order of increasing "end offsets". That way at least the emitted values are still always increasing.
While we can't really guarantee ordering with this new rescalable implementation, we could still do a best effort on that locally. What do you think?